### PR TITLE
Add support for single-species compara dbs

### PR DESCRIPTION
**Pull request by ens-ap5**
## Requirements

- Filling out the template is required. Any pull request that does not include enough information to be efficiently reviewed may be rejected.
- Please consider which branch this is to be submitted against. This is usually obvious, however if it is to be applied to both a release branch nn and master then please submit it against postreleasefix/nn and let us merge it into to the two branches.

## Description

Add support for single-species compara databases (and also remove some obsolete compara code).

## Views affected

Should only affect the rapid release site, enabling the configuration of the new Homologues view which is found in the ensembl-rapid repository.

## Possible complications

Unlikely, as changes are triggered by the presence of the following configuration in ensembl-rapid:

`$SiteDefs::SINGLE_SPECIES_COMPARA = 1`

## Merge conflicts

Unlikely - the changes are mostly to low-level code.

## Related JIRA Issues (EBI developers only)

https://www.ebi.ac.uk/panda/jira/browse/ENSWEB-6136

### CODE REVIEW

File: modules/EnsEMBL/Web/Object/Gene.pm [hunk 1]
```diff
@@ -787,16 +752,13 @@ sub get_homologies {
...
-  my %homologues;
-
   return unless $database;
 
-  my $query_member   = $database->get_GeneMemberAdaptor->fetch_by_stable_id($geneid);
-
+  my $query_member   = $database->get_GeneMemberAdaptor->fetch_by_stable_id($self->stable_id);
   return unless defined $query_member;
   
+  my %homologues;
```
Comment (ens-ma7) [inline modules/EnsEMBL/Web/Object/Gene.pm:761]: I am abit confused with this line, you removed the %homologues from line 792 which makes sense but why did you add it back here? It feels like its not being used elsewhere in that function so can be removed unless i missed something.

Comment (ens-ap5) [reply modules/EnsEMBL/Web/Object/Gene.pm:761]: You're right! I think I moved it down on the grounds that it's kinda pointless to initialise a variable just before a conditional return, then forgot to check if the variable was actually used. I'll remove it.

File: modules/EnsEMBL/Web/Query/Availability/Gene.pm [hunk 2]
```diff
@@ -44,6 +44,19 @@ sub fixup {
...
+sub _count_homologues {
+  my ($self, $member, $args) = @_;
+
+  my $dbc   = $self->database_dbc($args->{'species'},'compara');
+  my $gmID  = $member->gene_member_id;
+  my $sql   = 'select count(*) from homology_member where gene_member_id = ?';
+  my $sth   = $dbc->prepare($sql);
+  $sth->execute($gmID);
+
+  my @results = @{$sth->fetchall_arrayref||[]};
```
Comment (ens-ma7) [inline modules/EnsEMBL/Web/Query/Availability/Gene.pm:56]: Should this line be like this:
`@{$sth->fetchall_arrayref} ||[]` 

feels wrong to dereference an empty array. Also the return line below might produce an error if the array is empty.  can we check if this line is returning the right result?

File: modules/EnsEMBL/Web/QueryStore/Source/Adaptors.pm [hunk 3]
```diff
@@ -37,9 +37,15 @@ sub _database {
   my ($self,$species,$db) = @_;
 
   $db ||= 'core';
-  if($db =~ /compara/) {
-    $species = 'multi';
+
+  ## Rapid site has both single-species and multi-species compara
+  if ($self->{_sd}->SINGLE_SPECIES_COMPARA) {
```
Comment (ens-ma7) [inline modules/EnsEMBL/Web/QueryStore/Source/Adaptors.pm:42]: How can we test these changes? Can we see in a sandbox? 

Does this affect precache?

File: modules/EnsEMBL/Web/SpeciesDefs.pm [hunk 4]
```diff
@@ -631,6 +632,7 @@ sub _expand_database_templates {
   ## Autoconfigure databases
   unless (exists $tree->{'databases'} && exists $tree->{'databases'}{'DATABASE_CORE'}) {
     my @db_types = qw(CORE CDNA OTHERFEATURES RNASEQ FUNCGEN VARIATION);
+    push @db_types, 'COMPARA' if $SiteDefs::SINGLE_SPECIES_COMPARA;
```
Comment (ens-ma7) [inline modules/EnsEMBL/Web/SpeciesDefs.pm:635]: just checking, should the db_types be compara or pan_compara?

File: modules/EnsEMBL/Web/Query/Availability/Gene.pm (hunk 2 above)
Comment (ens-ap5) [reply modules/EnsEMBL/Web/Query/Availability/Gene.pm:56]: We should allow for a gene having no homologues, but you're right, that line will fail if the array is empty!

File: modules/EnsEMBL/Web/QueryStore/Source/Adaptors.pm (hunk 3 above)
Comment (ens-ap5) [reply modules/EnsEMBL/Web/QueryStore/Source/Adaptors.pm:42]: It's tricky to test right now, because I'm waiting for RR compara dbs to be handed over. The new code here will be ignored by all sites apart from RR, and the changes will be tested on RR before that can go live.

File: modules/EnsEMBL/Web/SpeciesDefs.pm (hunk 4 above)
Comment (ens-ap5) [reply modules/EnsEMBL/Web/SpeciesDefs.pm:635]: It should be compara - each species has a separate compara database, so it has to be processed with the other species-specific databases. There's also a separate "pan compara" database that contains reference species from www and the NV divisions.

File: modules/EnsEMBL/Web/QueryStore/Source/Adaptors.pm (hunk 3 above)
Comment (ens-ap5) [reply modules/EnsEMBL/Web/QueryStore/Source/Adaptors.pm:42]: So no, no effect on precache, because RR doesn't have it.

**Review by ens-ma7 — APPROVED** [review]

**Comment by ens-ma7** [comment]
I have gone through the code and approved it. I have mostly check the code change looks fine to my best knowledge, we couldn't check on a sandbox. Even though the changes should be for rapid release only, will be good to keep an eye on staging/test (rebuilding packed files for compara) and of course to check the changes are on rapid release (will be good to test on a staging site if there is one for rapid release).

**Comment by ens-st3** [comment]
merge into postreleasefix/104 ?

**Comment by ens-ap5** [comment]
> merge into postreleasefix/104 ?

I thought there might be too many conflicts with my recent changes to SpeciesDefs for an mpush to work smoothly, so it might be best to do it separately

**Comment by ens-ma7** [comment]
> > merge into postreleasefix/104 ?
> 
> I thought there might be too many conflicts with my recent changes to SpeciesDefs for an mpush to work smoothly, so it might be best to do it separately

I have a feeling this might cause problem with mpush with 104 as this will cause postreleasefix/104 branch to be behind.

**Comment by ens-ma7** [comment]
After our discussion in the code review, we suggested to push these commits to master once the release/105 branch is in. And then to use mpush to push the changes to both master and release/105

**Review by jyothishnt — APPROVED** [review]
yeah. Since it has been branched of release/104 its better to merge it to the same parent.

File: modules/EnsEMBL/Web/Object/Gene.pm (hunk 1 above)
[the remark this answers is not in the document]
Comment (ens-ap5) [reply modules/EnsEMBL/Web/Object/Gene.pm]: I deleted that variable in a commit 2 weeks ago.
